test/firestore_test.dart

```diff
@@ -421,6 +421,71 @@ void main() {
         expect(list.length, 1);
         expect(list.first.reference.documentID, "two");
       });
+
+      test('snapshots changes', () async {
+        var collRef =
+            app.firestore().collection('tests/query/snapshots_changes');
+
+        var docRef = collRef.document('item');
+        // delete it
+        await docRef.delete();
+
+        var completer1 = new Completer();
+        var completer2 = new Completer();
+        var completer3 = new Completer();
+        var completer4 = new Completer();
+        int count = 0;
+        var subscription =
+            collRef.snapshots.listen((QuerySnapshot querySnapshot) {
+          if (++count == 1) {
+            // first step ignore the result
+            completer1.complete();
+          } else if (count == 2) {
+            // second step expect an added item
+            expect(querySnapshot.documentChanges.length, 1);
+            expect(querySnapshot.documentChanges.first.type,
+                DocumentChangeType.added);
```

Comment:
nit: Expectations inside stream listener look a bit out of place. I'd suggest to make completers return `querySnapshot` (or even `querySnapshot.documentChanges`) and perform expectations after awaits. E.g.

```dart
var completer1 = new Completer<List<DocumentChange>>();
// the same other completers
var subscription = colRef.snapshots.listen((snapshot) {
  if (++count == 1) {
    completer1.complete(snapshot.documentChanges);
  } else if (count == 2) {
    completer2.complete(snapshot.documentChanges);
  } else {
    // et cetera.
  }
});

var changes1 = await completer1.future;
expect(changes1, isEmpty);
await docRef.setData(new DocumentData());
var changes2 = await completer2.future;
expect(changes2, hasLength(1));
expect(changes2.single.type, DocumentChangeType.added);
// et cetera.
```

This makes it a little bit easier to follow the test case flow, at least for me.